include/core/gate.hpp

```diff
@@ -0,0 +1,277 @@
+
+        /**
+     * @brief Default deconstructor
+     *
+     * Clears private data
+     *
+
+     */
+        virtual ~gate();
+
+        /**
+     * @brief Assignment operator
+     *
+     * @param other Gate to be assigned
+     *
+     * @return Pointer to instance
+     *
+
+     */
+        gate& operator=(const gate& other);
+
+        /**
+     * @brief Start iterator for accessing control lines. 
+     *
+     * Returns The start iterator of the line_container for accessing control lines. 
+     *
+
+
+     */
+        [[nodiscard]] virtual const_iterator begin_controls() const;
+
+        /**
+     * @brief End iterator for accessing control lines (const).
+     *
+     * Returns The end iterator of the line_container for accessing control lines (const).
+     *
+
+
+     */
+        [[nodiscard]] virtual const_iterator end_controls() const;
+
+        /**
+     * @brief Start iterator for accessing target lines (const).
+     *
+     * Returns The start iterator of the line_container for accessing target lines (const). 
+     *
+
+
+     */
+        [[nodiscard]] virtual const_iterator begin_targets() const;
+
+        /**
+     * @brief End iterator for accessing target lines (const).
+     *
+     * Returns The end iterator of the line_container for accessing target lines (const). 
+     *
+
+
+     */
+        [[nodiscard]] virtual const_iterator end_targets() const;
+
+        /**
+     * @brief Adds a control line to the gate
+     *
+     * @param c control line to add
+     *
+
+
+     */
+        virtual void add_control(line c);
+
+        /**
+     * @brief Adds a target to the desired line
+     *
+     * @param l target line 
+     *
+
+
+     */
+        virtual void add_target(line l);
+
+        /**
+     * @brief Sets the type of the target line(s)
+     *
+     * @param t target type
+     *
+
+     */
+        virtual void set_type(const std::any& t);
+
+        /**
+     * @brief Returns the type of the target line(s)
+     *
+     * @return target type
+     *
+
+     */
+        [[nodiscard]] virtual const std::any& type() const;
+
+    private:
+        struct priv;
+        priv* const d;
+    };
+
+    struct transform_line {
+        typedef gate::line result_type;
+
+        transform_line() = default;
+
+        explicit transform_line(const std::vector<unsigned>& filter):
+            filter(&filter) {}
+
+        gate::line operator()(gate::line l) const {
+            return filter ? std::find(filter->begin(), filter->end(), l) - filter->begin() : l;
+        }
+
+    private:
+        const std::vector<unsigned>* filter = nullptr;
+    };
+    /** @endcond */
+
+    /** @cond */
+    struct filter_line {
+        filter_line() = default;
+
+        explicit filter_line(const std::vector<unsigned>& filter):
+            filter(&filter) {}
+
+        bool operator()(const gate::line& l) const {
+            return !filter || std::find(filter->begin(), filter->end(), l) != filter->end();
+        }
```

Comment:
these are never actively used and introduce an annoying dependency on boost. 
Consider removing them and all the `transform_iterator` and `filter_iterator` stuff.